staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -305,6 +324,8 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) error {
 	go func() {
 		<-stopCh
 		close(internalStopCh)
+		s.HandlerChainWaitGroup.Wait()
```

Comment:
After moving everything over to our own WaitGroup (which looks all good btw code-wise!), one question from a few steps back:

The internalStopCh is eventually passed down to `RunServer`. `RunServer` will shutdown the http server below: `server.Shutdown()`. After the http server is down, no **new** request is going to be processed.

Q: does `server.Shutdown()` enforce that all request handlers have been returned?

About the race we saw in CI: the `s.HandlerChainWaitGroup.Wait()` here is potentially called before `server.Shutdown()` has returned. So even if the upper question is answered with 'yes', the `Wait` is called before the last `waitGroup.Add`. This is the race we saw (and this race is fixed with our own WaitGroup now).

Now my concern:

In the situation of the race we saw (i.e. Add after Wait at counter==0), this `s.HandlerChainWaitGroup.Wait()` with our new WaitGroup will return early before the last request has been processed. Is this true? If yes, we loose events, don't we?

In other words: I have the suspicion that the occurance of the detected race (with the old WaitGroup) means that we loose events now (with the new WaitGroup).

Reply:
>Q: does server.Shutdown() enforce that all request handlers have been returned?

For all the connections, Shutdown will wait until they turn into idle state,  then close them.  If a connection always stay active, Shutdown will close it until timeout.

>In other words: I have the suspicion that the occurance of the detected race (with the old WaitGroup) means that we loose events now (with the new WaitGroup).

In theory, events lost can exist. And it is with very low probability.